cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py

```diff
@@ -15,6 +17,38 @@
 from ._torchvision import TorchvisionFunction
 
 
+def _is_positively_oriented(contour: np.ndarray) -> bool:
+    ys, xs = contour.T
+
+    # This is the shoelace formula, except we only need the sign of the result,
+    # so we compare instead of subtracting. Compared to the typical formula,
+    # the sign is inverted, because the Y axis points downwards.
+    return np.sum(xs * np.roll(ys, -1)) < np.sum(ys * np.roll(xs, -1))
+
+
+def _generate_shapes(
+    context: cvataa.DetectionFunctionContext, box: Tensor, mask: Tensor, label: Tensor
+) -> Iterator[models.LabeledShapeRequest]:
+    LEVEL = 0.5
+
+    if context.conv_mask_to_poly:
+        # Since we treat mask values of exactly LEVEL as true, we'd like them
+        # to also be considered high by find_contours. And for that, the level
+        # parameter must be slightly less than LEVEL.
+        contours = measure.find_contours(mask[0].detach().numpy(), level=math.nextafter(LEVEL, 0))
+
+        for contour in contours:
+            if len(contour) < 3 or _is_positively_oriented(contour):
+                continue
+
+            contour = measure.approximate_polygon(contour, tolerance=2.5)
+
+            yield cvataa.polygon(label.item(), contour[:, ::-1].ravel().tolist())
```

Comment:
Probably, these polygons should be grouped, if there are more than 1 of them.

Reply:
As an option, assign a group, equal to the index, to each annotation, if no group is assigned already.

Reply:
I probably didn't describe the problem clearly enough: the problem is selecting group IDs that do not exist in the **remote** dataset. The user has the option to add the auto-generated annotations to existing annotations in the task, and it's important that new shapes are not added to existing groups.

As far as I can see, your suggestions only make sure that the auto-generated group IDs don't conflict with each other, which is good, but not sufficient.

I could solve this problem by fetching existing annotations, finding the maximum existing group ID and assigning new group IDs starting from that, but I don't really want to do that. It's very inefficient, and is susceptible to race conditions.

IMO, a proper solution lies beyond the scope of this PR and would involve server-side code. (E.g. the server could support special group values that would mean "assign a new group".)

Reply:
Ok, I thought the annotations are supposed to be overwritten in the result of the call. If not, then it indeed requires a more complicated solution.

Reply:
The overwrite is optional (it's controlled by the `clear_existing` flag). This mirrors the "Clean previous annotations" checkbox in the UI.